File: 50-mobile/52-mobile-evaluation/ios/52-mobile-evaluation/FlagSession.swift
import Combine
import Foundation
import LaunchDarkly

/// LaunchDarkly capability: iOS mobile SDK — start, string/bool variation,
/// observe, close.
/// https://launchdarkly.com/docs/sdk/client-side/ios
///
/// Credential is Info.plist LDMobileKey from LD_MOBILE_KEY in xcconfig.
final class FlagSession: ObservableObject {
    static let highlightKey = "enable-mobile-grid-highlight"
    static let countKey = "show-mobile-move-count"

    @Published var highlight: String = "none"
    @Published var showCount: Bool = false
    @Published var initializeCount: Int = 0
    @Published var closeCount: Int = 0
    @Published var changeHighlightCount: Int = 0
    @Published var changeCountFlagCount: Int = 0
    @Published var hasMobileKey: Bool = false
    @Published var status: String = "SDK not started"

    private static func mobileKey() -> String {
        (Bundle.main.object(forInfoDictionaryKey: "LDMobileKey") as? String ?? "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    func start(username: String) {
        let key = Self.mobileKey()
        hasMobileKey = !key.isEmpty
        guard !key.isEmpty else {
            highlight = "none"
            showCount = false
            status = "No LD_MOBILE_KEY — serving code defaults"
            return
        }
        status = "Initializing…"
        let config = LDConfig(mobileKey: key, autoEnvAttributes: .enabled)
        let builder = LDContextBuilder(key: username)
        guard case .success(let context) = builder.build() else {
            status = "Invalid context — serving code defaults"
            return
        }
        LDClient.start(config: config, context: context, startWaitSeconds: 5) { [weak self] _ in
            DispatchQueue.main.async {
                guard let self else { return }
                self.initializeCount += 1
                self.observe()
                self.readFlags()
                self.status = "Connected"
            }
        }
    }

    func stop() {
        if let client = LDClient.get() {
            client.stopObserving(owner: self)
            client.close()
            closeCount += 1
        }
        highlight = "none"
        showCount = false
        status = "Closed"
    }

    var sdkLog: String {
        "initialize ×\(initializeCount)\n" +
            "change:\(Self.highlightKey) ×\(changeHighlightCount)\n" +
            "change:\(Self.countKey) ×\(changeCountFlagCount)\n" +
            "close ×\(closeCount)"
    }

    private func observe() {
        LDClient.get()?.observe(keys: [Self.highlightKey, Self.countKey], owner: self) { [weak self] changed in
            DispatchQueue.main.async {
                guard let self else { return }
                if changed[Self.highlightKey] != nil {
                    self.changeHighlightCount += 1
                }
                if changed[Self.countKey] != nil {
                    self.changeCountFlagCount += 1
                }
                self.readFlags()
            }
        }
    }

    private func readFlags() {
        guard let client = LDClient.get() else { return }
        highlight = client.stringVariation(forKey: Self.highlightKey, defaultValue: "none")
        showCount = client.boolVariation(forKey: Self.countKey, defaultValue: false)
    }
}

File: 50-mobile/52-mobile-evaluation/android/app/src/main/java/com/lunchmarcoly/evaluation52/FlagSession.kt
package com.lunchmarcoly.evaluation52

import android.app.Application
import android.os.Handler
import android.os.Looper
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.launchdarkly.sdk.LDContext
import com.launchdarkly.sdk.android.FeatureFlagChangeListener
import com.launchdarkly.sdk.android.LDClient
import com.launchdarkly.sdk.android.LDConfig

/**
 * LaunchDarkly capability: Android mobile SDK — initialize, string/boolean
 * variation, flag listeners, close.
 * https://launchdarkly.com/docs/sdk/client-side/android
 *
 * Credential is BuildConfig.LD_MOBILE_KEY (from local.properties ld.mobile.key
 * or the LD_MOBILE_KEY env var). Never an SDK key.
 */
class FlagSession {
    var highlight by mutableStateOf("none")
        private set
    var showCount by mutableStateOf(false)
        private set
    var initializeCount by mutableIntStateOf(0)
        private set
    var closeCount by mutableIntStateOf(0)
        private set
    var changeHighlightCount by mutableIntStateOf(0)
        private set
    var changeCountFlagCount by mutableIntStateOf(0)
        private set
    var hasMobileKey by mutableStateOf(false)
        private set
    var status by mutableStateOf("SDK not started")
        private set

    private val main = Handler(Looper.getMainLooper())
    private var client: LDClient? = null

    private val highlightListener = FeatureFlagChangeListener {
        main.post {
            changeHighlightCount += 1
            readFlags()
        }
    }

    private val countListener = FeatureFlagChangeListener {
        main.post {
            changeCountFlagCount += 1
            readFlags()
        }
    }

    fun start(app: Application, username: String) {
        val key = BuildConfig.LD_MOBILE_KEY.trim()
        hasMobileKey = key.isNotEmpty()
        if (key.isEmpty()) {
            highlight = "none"
            showCount = false
            status = "No LD_MOBILE_KEY — serving code defaults"
            return
        }
        status = "Initializing…"
        Thread {
            try {
                val config = LDConfig.Builder(LDConfig.Builder.AutoEnvAttributes.Enabled)
                    .mobileKey(key)
                    .build()
                val context = LDContext.create(username)
                val started = LDClient.init(app, config, context, 5)
                main.post {
                    client = started
                    initializeCount += 1
                    started.registerFeatureFlagListener(HIGHLIGHT_KEY, highlightListener)
                    started.registerFeatureFlagListener(COUNT_KEY, countListener)
                    readFlags()
                    status = "Connected"
                }
            } catch (e: Exception) {
                main.post {
                    highlight = "none"
                    showCount = false
                    status = "Init failed — serving code defaults"
                }
            }
        }.start()
    }

    fun stop() {
        val c = client ?: return
        try {
            c.unregisterFeatureFlagListener(HIGHLIGHT_KEY, highlightListener)
            c.unregisterFeatureFlagListener(COUNT_KEY, countListener)
            c.close()
        } catch (_: Exception) {
        }
        client = null
        closeCount += 1
        highlight = "none"
        showCount = false
        status = "Closed"
    }

    fun sdkLog(): String =
        "initialize ×$initializeCount\n" +
            "change:$HIGHLIGHT_KEY ×$changeHighlightCount\n" +
            "change:$COUNT_KEY ×$changeCountFlagCount\n" +
            "close ×$closeCount"

    private fun readFlags() {
        val c = client ?: return
        highlight = c.stringVariation(HIGHLIGHT_KEY, "none")
        showCount = c.boolVariation(COUNT_KEY, false)
    }

    companion object {
        const val HIGHLIGHT_KEY = "enable-mobile-grid-highlight"
        const val COUNT_KEY = "show-mobile-move-count"
    }
}
